# Web Scraping and Cleaning: USGS Monitor Location Analysis

## 1.) Install Beautiful Soup

In [1]:
%pip install beautifulsoup4

Note: you may need to restart the kernel to use updated packages.


## 2.) Conduct WebScraping

In [8]:
import requests 
from bs4 import BeautifulSoup
import pandas as pd

url= "https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?limit=2000&f=html"
response= requests.get(url)
soup= BeautifulSoup(response.text, 'html.parser')

In [9]:
response.status_code

200

In [10]:
divs = soup.find_all("div", class_="usa-table-container")
len(divs)

1

In [12]:
table = divs[0].find("table")

In [13]:
table is not None

True

In [15]:
headerlist=[]
datalist=[];
divs = soup.find_all("div", class_="usa-table-container")

for div in divs:

  for table in div.find_all("table", class_= "usa-table usa-table--striped"):
    for headrow in table.find_all("thead"):
      headers = headrow.find_all("tr") ;
      for header in headers:
          head=header.find_all("th");
          for h in head:
            headerlist.append(h.text.strip());

In [16]:
headerlist

['id',
 'agency_code',
 'agency_name',
 'monitoring_location_number',
 'monitoring_location_name',
 'district_code',
 'country_code',
 'country_name',
 'state_code',
 'state_name',
 'county_code',
 'county_name',
 'minor_civil_division_code',
 'site_type_code',
 'site_type',
 'hydrologic_unit_code',
 'basin_code',
 'altitude',
 'altitude_accuracy',
 'altitude_method_code',
 'altitude_method_name',
 'vertical_datum',
 'vertical_datum_name',
 'horizontal_positional_accuracy_code',
 'horizontal_positional_accuracy',
 'horizontal_position_method_code',
 'horizontal_position_method_name',
 'original_horizontal_datum',
 'original_horizontal_datum_name',
 'drainage_area',
 'contributing_drainage_area',
 'time_zone_abbreviation',
 'uses_daylight_savings',
 'construction_date',
 'aquifer_code',
 'national_aquifer_code',
 'aquifer_type_code',
 'well_constructed_depth',
 'hole_constructed_depth',
 'depth_source_code',
 'revision_note',
 'revision_created',
 'revision_modified']

In [17]:
for datarow in table.find_all("tbody"):
      rows = datarow.find_all("tr") ;
      for row in rows:
          data=row.find_all("td");
          rowlist=[]
          for d in data:
            rowlist.append(d.text.strip());
          datalist.append(rowlist);

In [18]:
len(datalist)

2000

Check

In [21]:
datalist[:3]

[['AL012-90100100001',
  'AL012',
  'Alabama Office of Water Resources',
  '90100100001',
  'Autauga County Water Authority',
  '01',
  'US',
  'United States of America',
  '01',
  'Alabama',
  '001',
  'Autauga County',
  'None',
  'FA-WDS',
  'Water-distribution system',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'CST',
  'Y',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None'],
 ['AL012-90100100002',
  'AL012',
  'Alabama Office of Water Resources',
  '90100100002',
  'Autaugaville Water System',
  '01',
  'US',
  'United States of America',
  '01',
  'Alabama',
  '001',
  'Autauga County',
  'None',
  'FA-WDS',
  'Water-distribution system',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'CST',
  'Y',
  

In [23]:
 df = pd.DataFrame(datalist, columns=headerlist)

In [24]:
df.head()

,id,agency_code,agency_name,monitoring_location_number,monitoring_location_name,district_code,country_code,country_name,state_code,state_name,...,construction_date,aquifer_code,national_aquifer_code,aquifer_type_code,well_constructed_depth,hole_constructed_depth,depth_source_code,revision_note,revision_created,revision_modified
0,AL012-90100100001,AL012,Alabama Office of Water Resources,90100100001,Autauga County Water Authority,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
1,AL012-90100100002,AL012,Alabama Office of Water Resources,90100100002,Autaugaville Water System,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
2,AL012-90100100003,AL012,Alabama Office of Water Resources,90100100003,Billingsley Water System,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
3,AL012-90100100005,AL012,Alabama Office of Water Resources,90100100005,Water Works Board Of Prattville,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
4,AL012-90100100006,AL012,Alabama Office of Water Resources,90100100006,West Autauga Water Authority,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None


In [25]:
df.shape

(2000, 43)

In [32]:
import time

In [33]:
url = "https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?limit=2000"

response = requests.get(url)
data = response.json()

In [35]:
data["links"]

[{'rel': 'next',
  'href': 'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?cursor=QVIwMDgtMzMxMjE0MDkyMzc1MzAx&limit=2000',
  'type': 'application/geo+json',
  'title': 'Items (next)'},
 {'type': 'application/geo+json',
  'rel': 'self',
  'title': 'This document as GeoJSON',
  'href': 'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?f=json&limit=2000'},
 {'rel': 'alternate',
  'type': 'application/ld+json',
  'title': 'This document as RDF (JSON-LD)',
  'href': 'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?f=jsonld&limit=2000'},
 {'type': 'text/html',
  'rel': 'alternate',
  'title': 'This document as HTML',
  'href': 'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?f=html&limit=2000'},
 {'type': 'text/csv; charset=utf-8',
  'rel': 'alternate',
  'title': 'This document as CSV',
  'href': 'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-loc

In [36]:
next_url = next(
    (link["href"] for link in data.get("links", [])
     if link.get("rel") == "next"),
    None
)

next_url

'https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?cursor=QVIwMDgtMzMxMjE0MDkyMzc1MzAx&limit=2000'

In [37]:
page2_url = next_url + "&f=html"

response2 = requests.get(page2_url)
soup2 = BeautifulSoup(response2.text, "html.parser")

In [38]:
response2.status_code

200

In [39]:
divs2 = soup2.find_all("div", class_="usa-table-container")
len(divs2)

1

In [40]:
headerlist = []
datalist = []

# Start with the first page in JSON format
url = "https://api.waterdata.usgs.gov/ogcapi/v1/collections/monitoring-locations/items?limit=2000"

for page in range(25):

    # Request JSON so we can get the next-page link
    response = requests.get(url)
    data = response.json()

    # Convert the same page to HTML for BeautifulSoup
    html_url = url + "&f=html"
    html_response = requests.get(html_url)
    soup = BeautifulSoup(html_response.text, "html.parser")

    divs = soup.find_all("div", class_="usa-table-container")

    for div in divs:

        for table in div.find_all(
            "table",
            class_="usa-table usa-table--striped"
        ):

            # Collect column headers only once
            if len(headerlist) == 0:

                for headrow in table.find_all("thead"):
                    headers = headrow.find_all("tr")

                    for header in headers:
                        head = header.find_all("th")

                        for h in head:
                            headerlist.append(h.text.strip())

            # Collect data
            for datarow in table.find_all("tbody"):
                rows = datarow.find_all("tr")

                for row in rows:
                    data_cells = row.find_all("td")
                    rowlist = []

                    for d in data_cells:
                        rowlist.append(d.text.strip())

                    datalist.append(rowlist)

    print(
        "Page", page + 1,
        "complete - Total records:",
        len(datalist)
    )

    # Get next page URL from JSON
    next_url = next(
        (
            link["href"]
            for link in data.get("links", [])
            if link.get("rel") == "next"
        ),
        None
    )

    # Stop if there is no next page
    if next_url is None:
        print("No next page found.")
        break

    url = next_url

    time.sleep(0.5)

Page 1 complete - Total records: 2000
Page 2 complete - Total records: 4000
Page 3 complete - Total records: 6000
Page 4 complete - Total records: 8000
Page 5 complete - Total records: 10000
Page 6 complete - Total records: 12000
Page 7 complete - Total records: 14000
Page 8 complete - Total records: 16000
Page 9 complete - Total records: 18000
Page 10 complete - Total records: 20000
Page 11 complete - Total records: 22000
Page 12 complete - Total records: 24000
Page 13 complete - Total records: 26000
Page 14 complete - Total records: 28000
Page 15 complete - Total records: 30000
Page 16 complete - Total records: 32000
Page 17 complete - Total records: 34000
Page 18 complete - Total records: 36000
Page 19 complete - Total records: 38000
Page 20 complete - Total records: 40000
Page 21 complete - Total records: 42000
Page 22 complete - Total records: 44000
Page 23 complete - Total records: 46000
Page 24 complete - Total records: 48000
Page 25 complete - Total records: 50000


In [41]:
len(datalist)

50000

In [43]:
df = pd.DataFrame(datalist, columns=headerlist)

In [44]:
df.shape

(50000, 43)

In [45]:
df.head()

,id,agency_code,agency_name,monitoring_location_number,monitoring_location_name,district_code,country_code,country_name,state_code,state_name,...,construction_date,aquifer_code,national_aquifer_code,aquifer_type_code,well_constructed_depth,hole_constructed_depth,depth_source_code,revision_note,revision_created,revision_modified
0,AL012-90100100001,AL012,Alabama Office of Water Resources,90100100001,Autauga County Water Authority,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
1,AL012-90100100002,AL012,Alabama Office of Water Resources,90100100002,Autaugaville Water System,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
2,AL012-90100100003,AL012,Alabama Office of Water Resources,90100100003,Billingsley Water System,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
3,AL012-90100100005,AL012,Alabama Office of Water Resources,90100100005,Water Works Board Of Prattville,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None
4,AL012-90100100006,AL012,Alabama Office of Water Resources,90100100006,West Autauga Water Authority,01,US,United States of America,01,Alabama,...,None,None,None,None,None,None,None,None,None,None


## 3.) Data Cleanup

### The scraped USGS data is examined for missing values, duplicate records, and incorrect data types before statistical analysis.

In [49]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 43 columns):
 #   Column                               Non-Null Count  Dtype
---  ------                               --------------  -----
 0   id                                   50000 non-null  str  
 1   agency_code                          50000 non-null  str  
 2   agency_name                          50000 non-null  str  
 3   monitoring_location_number           50000 non-null  str  
 4   monitoring_location_name             50000 non-null  str  
 5   district_code                        50000 non-null  str  
 6   country_code                         50000 non-null  str  
 7   country_name                         50000 non-null  str  
 8   state_code                           50000 non-null  str  
 9   state_name                           50000 non-null  str  
 10  county_code                          50000 non-null  str  
 11  county_name                          50000 non-null  str  
 12  m

In [50]:
df.isnull().sum()

id                                     0
agency_code                            0
agency_name                            0
monitoring_location_number             0
monitoring_location_name               0
district_code                          0
country_code                           0
country_name                           0
state_code                             0
state_name                             0
county_code                            0
county_name                            0
minor_civil_division_code              0
site_type_code                         0
site_type                              0
hydrologic_unit_code                   0
basin_code                             0
altitude                               0
altitude_accuracy                      0
altitude_method_code                   0
altitude_method_name                   0
vertical_datum                         0
vertical_datum_name                    0
horizontal_positional_accuracy_code    0
horizontal_posit

In [76]:
df.replace("None", pd.NA, inplace=True)

,id,agency_code,agency_name,monitoring_location_number,monitoring_location_name,district_code,country_code,country_name,state_code,state_name,...,construction_date,aquifer_code,national_aquifer_code,aquifer_type_code,well_constructed_depth,hole_constructed_depth,depth_source_code,revision_note,revision_created,revision_modified
0,AL012-90100100001,AL012,Alabama Office of Water Resources,90100100001,Autauga County Water Authority,01,US,United States of America,01,Alabama,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,AL012-90100100002,AL012,Alabama Office of Water Resources,90100100002,Autaugaville Water System,01,US,United States of America,01,Alabama,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,AL012-90100100003,AL012,Alabama Office of Water Resources,90100100003,Billingsley Water System,01,US,United States of America,01,Alabama,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,AL012-90100100005,AL012,Alabama Office of Water Resources,90100100005,Water Works Board Of Prattville,01,US,United States of America,01,Alabama,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,AL012-90100100006,AL012,Alabama Office of Water Resources,90100100006,West Autauga Water Authority,01,US,United States of America,01,Alabama,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,AR008-361...,AR008,Arkansas Natural Resources Commission,361146090421001,18N03E13DD1,05,US,United States of America,05,Arkansas,...,NaN,110ALVM,N100MSRVVL,NaN,120,NaN,O,NaN,NaN,NaN
49996,AR008-361...,AR008,Arkansas Natural Resources Commission,361146090442800,18N03E15DC1,05,US,United States of America,05,Arkansas,...,NaN,110ALVM,N100MSRVVL,NaN,100,NaN,NaN,NaN,NaN,NaN
49997,AR008-361...,AR008,Arkansas Natural Resources Commission,361146090491601,18N01E23BD1,05,US,United States of America,05,Arkansas,...,NaN,NaN,NaN,NaN,112,NaN,O,NaN,NaN,NaN
49998,AR008-361...,AR008,Arkansas Natural Resources Commission,361146090505701,18N02E15DC1,05,US,United States of America,05,Arkansas,...,NaN,110ALVM,N100MSRVVL,NaN,100,NaN,NaN,NaN,NaN,NaN


Missing values in the scraped dataset were represented by the text "None". These values were converted to NA so that Python could correctly recognize them as missing data.

### Check missing values again

In [77]:
df.isnull().sum()

id                                         0
agency_code                                0
agency_name                                0
monitoring_location_number                 0
monitoring_location_name                   0
district_code                              0
country_code                               0
country_name                               0
state_code                                 0
state_name                                 0
county_code                                0
county_name                                0
minor_civil_division_code              49533
site_type_code                             0
site_type                                  0
hydrologic_unit_code                     307
basin_code                             49971
altitude                               37621
altitude_accuracy                      37621
altitude_method_code                   37621
altitude_method_name                   37621
vertical_datum                         37621
vertical_d

### Checking duplicates

In [78]:
df.duplicated().sum()

np.int64(0)

No duplicate rows were found in the dataset

### Taking a look at the data types

In [79]:
df.dtypes

id                                         str
agency_code                                str
agency_name                                str
monitoring_location_number                 str
monitoring_location_name                   str
district_code                              str
country_code                               str
country_name                               str
state_code                                 str
state_name                                 str
county_code                                str
county_name                                str
minor_civil_division_code                  str
site_type_code                             str
site_type                                  str
hydrologic_unit_code                       str
basin_code                                 str
altitude                               float64
altitude_accuracy                      float64
altitude_method_code                       str
altitude_method_name                       str
vertical_datu

### Convert columns needed for analysis

In [80]:
df["altitude"] = pd.to_numeric(df["altitude"], errors="coerce")
df["altitude_accuracy"] = pd.to_numeric(df["altitude_accuracy"], errors="coerce")

In [81]:
df[["altitude", "altitude_accuracy"]].dtypes

altitude             float64
altitude_accuracy    float64
dtype: object

The scraped dataset initially stored all columns as strings. The altitude and altitude_accuracy columns were converted to numeric values because they are needed for the statistical analysis.

## 4.) Statistical Analysis

###     I.) How are USGS monitoring locations distributed across different states?

Frequency:

In [82]:
state_frequency = df["state_name"].value_counts()
state_frequency

state_name
Arkansas    49694
Alabama       306
Name: count, dtype: int64

Percentage:

In [83]:
state_percentage = df["state_name"].value_counts(normalize=True) * 100
state_percentage.round(2)

state_name
Arkansas    99.39
Alabama      0.61
Name: proportion, dtype: float64

Mode:

In [84]:
df["state_name"].mode()

0    Arkansas
Name: state_name, dtype: str

Of the 50,000 USGS monitoring locations scraped, Arkansas contained 49,694 locations, representing approximately 99.39% of the dataset. Alabama contained 306 locations, representing approximately 0.61%. Arkansas was therefore the mode, meaning it was the most frequently occurring state in the scraped dataset.

### II.) How does the distribution of monitoring-location altitude differ among states?

In [85]:
altitude_by_state = df.groupby("state_name")["altitude"].agg(
    Mean="mean",
    Median="median",
    Minimum="min",
    Maximum="max"
)

altitude_by_state

,Mean,Median,Minimum,Maximum
state_name,,,,
Alabama,NaN,NaN,NaN,NaN
Arkansas,206.882243,204.0,0.0,1600.0


For Arkansas, there are valid numeric altitude values, so Python calculated:
- Mean: 206.88
- Median: 204
- Minimum: 0
- Maximum: 1600
  
For Alabama, every result is NaN. Meaning the 306 Alabama records do not have usable altitude values. 

In [87]:
df.groupby("state_name")["altitude"].count()

state_name
Alabama         0
Arkansas    12379
Name: altitude, dtype: int64

Alabama can't be included in the altitude comparison because it has 0 valid altitude values, while Arkansas has 12,379 valid altitude measurements

In [88]:
altitude_by_state["Q1"] = df.groupby("state_name")["altitude"].quantile(0.25)
altitude_by_state["Q3"] = df.groupby("state_name")["altitude"].quantile(0.75)
altitude_by_state["IQR"] = altitude_by_state["Q3"] - altitude_by_state["Q1"]

altitude_by_state.round(2)

,Mean,Median,Minimum,Maximum,Q1,Q3,IQR
state_name,,,,,,,
Alabama,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Arkansas,206.88,204.0,0.0,1600.0,187.0,220.0,33.0


Final Interpretation:

Alabama had no valid altitude values in the scraped dataset, so altitude statistics couldn't be calculated for that state. Arkansas had 12,379 valid altitude measurements. The mean altitude was 206.88, and the median was 204.00. Altitude values ranged from 0 to 1,600. The first quartile was 187.00 and the third quartile was 220.00, giving an IQR of 33.00. This means the middle 50% of Arkansas monitoring-location altitudes fell within a relatively narrow 33-unit range.

### III.) How much does altitude vary among the monitoring locations? 

In [69]:
altitude = df["altitude"].dropna()

altitude_range = altitude.max() - altitude.min()
altitude_std = altitude.std()
altitude_variance = altitude.var()

Q1 = altitude.quantile(0.25)
Q3 = altitude.quantile(0.75)
altitude_iqr = Q3 - Q1

In [72]:
print("Range:", round(altitude_range, 2))
print("Standard Deviation:", round(altitude_std, 2))
print("Variance:", round(altitude_variance, 2))
print("IQR:", round(altitude_iqr, 2))

Range: 1600.0
Standard Deviation: 53.07
Variance: 2816.66
IQR: 33.0


The altitude measurements show a wide overall spread. The range is 1,600, the standard deviation is 53.07, the variance is 2,816.66, and the IQR is 33.0.

### IV.) Which state has the lowest average altitude measurement error (highest average altitude accuracy)?

In [73]:
average_error = df.groupby("state_name")["altitude_accuracy"].mean()
average_error

state_name
Alabama          NaN
Arkansas    1.809902
Name: altitude_accuracy, dtype: float64

In [74]:
average_error.idxmin()

'Arkansas'

In [75]:
average_error.min()

np.float64(1.8099022538169482)

Arkansas had the lowest average altitude measurement error at approximately 1.81, which means it had the highest average altitude accuracy among the states with available data. Alabama had no valid altitude accuracy values.